Userland/Applications/PixelPaint/CloneTool.cpp

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (c) 2021, Mustafa Quraish <mustafa@cs.toronto.edu>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "CloneTool.h"
+#include "ImageEditor.h"
+#include "Layer.h"
+#include <LibGUI/Action.h>
+#include <LibGUI/BoxLayout.h>
+#include <LibGUI/CheckBox.h>
+#include <LibGUI/Label.h>
+#include <LibGUI/Menu.h>
+#include <LibGUI/Painter.h>
+#include <LibGUI/RadioButton.h>
+#include <LibGUI/ValueSlider.h>
+#include <LibGfx/Bitmap.h>
+
+namespace PixelPaint {
+
+CloneTool::CloneTool()
+{
+}
+
+CloneTool::~CloneTool()
+{
+}
+
+void CloneTool::draw_point(Gfx::Bitmap& bitmap, Gfx::Color const&, Gfx::IntPoint const& point)
+{
+    if (!m_sample_location.has_value())
+        return;
+
+    auto source_point = point - m_cursor_offset.value();
+    for (int y = -size(); y < size(); y++) {
+        for (int x = -size(); x < size(); x++) {
+            auto target_x = point.x() + x;
+            auto target_y = point.y() + y;
+            auto distance = point.distance_from({ target_x, target_y });
+            if (target_x < 0 || target_x >= bitmap.width() || target_y < 0 || target_y >= bitmap.height())
+                continue;
+            if (distance >= size())
+                continue;
+
+            auto source_x = source_point.x() + x;
+            auto source_y = source_point.y() + y;
+            if (source_x < 0 || source_x >= bitmap.width() || source_y < 0 || source_y >= bitmap.height())
+                continue;
+
+            auto falloff = (1.0 - double { distance / size() }) * (1.0 / (100 - hardness()));
+            auto pixel_color = bitmap.get_pixel(source_x, source_y);
+            pixel_color.set_alpha(falloff * 255);
+            bitmap.set_pixel(target_x, target_y, bitmap.get_pixel(target_x, target_y).blend(pixel_color));
+        }
+    }
+}
+
+void CloneTool::draw_line(Gfx::Bitmap& bitmap, Gfx::Color const& color, Gfx::IntPoint const& start, Gfx::IntPoint const& end)
+{
+    if (!m_sample_location.has_value())
+        return;
+    BrushTool::draw_line(bitmap, color, start, end);
+}
+
+void CloneTool::on_mousemove(Layer* layer, MouseEvent& event)
+{
+    if (m_is_selecting_location)
+        return;
+
+    if (m_cursor_offset.has_value()) {
+        m_sample_location = event.image_event().position() - m_cursor_offset.value();
+        // FIXME: This is a really inefficient way to update the marker's location
+        m_editor->update();
+    }
+
+    BrushTool::on_mousemove(layer, event);
+}
+
+void CloneTool::on_mousedown(Layer* layer, MouseEvent& event)
+{
+    if (m_is_selecting_location) {
```

Comment:
You have `event.alt()` which does exactly what you want here. It checks whether the modifier is pressed.
That way you can get rid of all the `m_is_selecting_location` business.